engines/lab/music.cpp

```diff
@@ -0,0 +1,339 @@
+/* ScummVM - Graphic Adventure Engine
+ *
+ * ScummVM is the legal property of its developers, whose names
+ * are too numerous to list here. Please refer to the COPYRIGHT
+ * file distributed with this source distribution.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+/*
+ * This code is based on Labyrinth of Time code with assistance of
+ *
+ * Copyright (c) 1993 Terra Nova Development
+ * Copyright (c) 2004 The Wyrmkeep Entertainment Co.
+ *
+ */
+
+#include "audio/decoders/raw.h"
+
+#include "lab/lab.h"
+
+#include "lab/anim.h"
+#include "lab/eventman.h"
+#include "lab/music.h"
+#include "lab/resource.h"
+
+namespace Lab {
+
+#define MUSICBUFSIZE   (2 * 65536L)
+#define SAMPLESPEED    15000L
+
+#define CLOWNROOM           123
+#define DIMROOM              80
+
+Music::Music(LabEngine *vm) : _vm(vm) {
+	_file = 0;
+	_tFile = 0;
+	_musicPaused = false;
+
+	_oldMusicOn = false;
+	_tLeftInFile = 0;
+
+	_leftInFile = 0;
+
+	_musicOn = false;
+	_loopSoundEffect = false;
+	_queuingAudioStream = nullptr;
+	_lastMusicRoom = 1;
+	_doReset = true;
+	_waitTillFinished = false;
+}
+
+void Music::updateMusic() {
```

Comment:
`updateMusic` is very poorly named, since it does event processing and screen updating as well. It might be more clear to move this function to the main engine class, and then keep only the actual music updating in the `Music` class.